src/openms/source/ANALYSIS/OPENSWATH/OpenSwathWorkflow.cpp

```diff
@@ -496,21 +503,47 @@ namespace OpenMS
             batch_size = batchSize;
           }
 
+          SignedSize nr_batches = (transition_exp_used_all.getCompounds().size() / batch_size);
 #ifdef _OPENMP
-#pragma omp critical (featureFinder)
+          // If we have a multiple of threads_outer_loop_ here, then use nested
+          // parallelization here. E.g. if we use 8 threads for the outer loop,
+          // but we have a total of 24 cores available, each of the 8 threads
+          // will then create a team of 3 threads to work on the batches
+          // individually.
+          //
+          // We should avoid oversubscribing the CPUs, therefore we use integer division.
+          // -- see https://docs.oracle.com/cd/E19059-01/stud.10/819-0501/2_nested.html
+          int outer_thread_nr = omp_get_thread_num();
+          omp_set_num_threads(std::max(1, total_nr_threads / threads_outer_loop_) );
```

Comment:
In a different context, I saw the rule not to use more than #cores - 1 threads (except if hyperthreading).
We do allow for #cores thread here right?

Reply:
well the user specifies the number of threads anyways. So in principle the user can set the number of threads to 128 on a single threaded machine and would get really poor performance. I am not sure whether there is an easy way to control for that if the user makes an error, we could check the maximum threads available but on a cluster where multiple jobs run per node that is not a good idea either (e.g. you have 3 jobs on a node with 48 cores, you get assigned 10 of these but your code doesnt know that it cannot use the other 38 and it looks like you have 48 for yourself). So in all these cases, the user needs to know and control how many cores to use. 

We should anyways test this on some real data, I am curious, where did you see the advise to use `#cores -1` threads?

Reply:
Mostly anecdotal ... would not give too much on that. I think it is intended to leave one core to the OS. Best is to benchmark I guess.

Reply:
I think that may also differ on Desktop machines that need to run an X server and other stuff vs HPC nodes that are dedicated for compute.